SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslationManager.cpp

```diff
@@ -47,25 +48,40 @@ void SpatialVirtualWorkerTranslationManager::AuthorityChanged(const Worker_Compo
 
 	if (!bAuthoritative)
 	{
+		// A healthy server should never lose auth (we supported worker recovery, but that only happens for disconnected servers).
 		UE_LOG(LogSpatialVirtualWorkerTranslationManager, Error,
-			   TEXT("Lost authority over the translation mapping. This is not supported."));
+			   TEXT("Lost authority over the translation mapping. This should never happen."));
 		return;
 	}
 
 	const int32 ExistingTranslatorMappingCount = Translator->GetMappingCount();
+	// Fresh deployment. We need to create partition entities before we start assigning virtual workers.
 	if (ExistingTranslatorMappingCount == 0)
 	{
-		// Fresh deployment, we need to create partition entities before we start assigning virtual workers.
 		UE_LOG(LogSpatialVirtualWorkerTranslationManager, Log,
 			   TEXT("Gained authority over the VirtualWorker translation, spawning partition entities."));
 		SpawnPartitionEntitiesForVirtualWorkerIds();
 	}
+	// Previously snapshot partition auth server crashed. Need to find that server's entry in the translator mapping
+	// so we're ready to reassign the virtual worker when the replacement server starts.
+	else if (Translator->IsReady())
+	{
+		// Because we've this server has just gained authority, the local VTM state will be empty, so we copy over
+		// from the translator.
+		VirtualToPhysicalWorkerMapping = Translator->VirtualToPhysicalWorkerMapping;
+
+		UE_LOG(LogSpatialVirtualWorkerTranslationManager, Warning,
+			   TEXT("Newly elected VTM server received authority. Cleaning up translator map."));
+		CleanupTranslatorMappingAfterAuthorityChange();
+	}
+	// Restarting from snapshot: reusing partition entities but have a whole new set of partition entities.
 	else if (ExistingTranslatorMappingCount == NumVirtualWorkers)
 	{
 		// Partitions already exist, reclaim them with latest server worker entities
 		UE_LOG(LogSpatialVirtualWorkerTranslationManager, Log,
-			   TEXT("Gained authority over the VirtualWorker translation, reclaiming partition entities."));
-		ReclaimPartitionEntities();
+			   TEXT("Gained authority over the VirtualWorker translation after snapshot restart. Resetting virtual worker mapping."));
+		ResetVirtualWorkerMappingAfterSnapshotReset();
+		QueryForServerWorkerEntities();
```

Comment:
Hmmm I'm not sure why we're querying for server worker entities, as we always have interest now. I must have failed to clean this up in the v15 upgrade.

Reply:
This is still in line 100 of the original file, I just moved it out of the Reset method because it seemed wrong to put it inside

Reply:
Yep - to my second sentence there - `I must have failed to clean this up in the v15 upgrade.`

Reply:
will create a ticket to clean up